src/module.model/poolpair.ts

```diff
@@ -0,0 +1,62 @@
+import { Injectable } from '@nestjs/common'
+import { Model, ModelMapping } from '@src/module.database/model'
+import { Database, SortOrder } from '@src/module.database/database'
+
+const PoolPairMapping: ModelMapping<PoolPair> = {
+  type: 'poolpair',
+  index: {
+    poolId: {
+      name: 'poolpair_poolId',
+      partition: {
+        type: 'string',
+        key: (p: PoolPair) => p.poolId
+      }
+    }
```

Comment:
Little confused about this index purpose, shouldn't it be

```suggestion
    poolId: {
      name: 'poolpair_poolId_height',
      partition: {
        type: 'string',
        key: (p: PoolPair) => p.poolId
      },
      sort: {
        type: 'height',
        key: (p: PoolPair) => p.block.height
      }
    }
```

for you to `getLatest(poolId)` information about the poolpair

Reply:
because the usecase i was trying to make the query do this way
```ts
const poolpairs = await mapper.query() // list

const poolpair = await mapper.get('3') // get

const latest = await mapper.getLatest() // return latest poolpair
```
its a very simple design like block
so if set `sort` which the query will become
```ts
const poolpairs = await mapper.query('3') // return ??
```

Reply:
For the PoolPair indexes by right you should not be tracking `PoolPair` you should be tracking `PoolPairMutations` changes.

This means you track changes within PoolPair as it processes within the blockchain. `getLatest('poolId')` allow you to get the latest mutated change of that PoolPair to use for calculation of something else.

Reply:
hmm.. @fuxingloh
In order to adopt poolswap index, the sources only container `fromTokenId` and `toTokenId`
so the poolpair index id will be constructed by tokenA-tokenB
while tracking `mutations`, then will have to store diff poolpair id, perhaps construct the id by tokenA-tokenB-{something}
then issue comes back to poolswap index, you got idea to get poolpair by id?? 

https://github.com/DeFiCh/whale/blob/539e9bed73c0736d908335a2f95146fd3f54c9d9/src/module.indexer/model/poolswap.aggregation.ts#L29-L42

Reply:
That would be `a-b-height` since height is mutating upwards.

You don't need to know something, you just need to get `a-b-` hence using getLatest() or getHighest()

For example `script.aggregation`, it uses the latest to get previous and build on top of it. This works because the ledger is always moving forward and getLatest is always the most recent.

https://github.com/DeFiCh/whale/blob/16d0f9813705dab2d9297f8301ee610c46dc3567/src/module.indexer/model/script.aggregation.ts#L58-L72

Reply:
hmm. i put poolId as partition key currently, if that case then
```ts
// from
model {
  id: tokenAId-tokenBId // unique
  poolId: '2', // partition
  height: 22, // sort
}

// to
model {
  id: tokenAId-tokenBId-height
  symbolId:  tokenAId-tokenBId // partition
  height // sort
  poolId
}
```

Reply:
yup, this works, you can choose to use `poolId` or `tokenAId-tokenBId` 

depending on your use case, I think yours is `require getLatest(tokenAId-tokenBId)`?

Reply:
yes use tokenAId-tokenBId is more straight forward due to poolswap object design, the current implementation is kind of tweaking around the workflow to set poolId as partitionKey :sweat_smile: